vpr/src/pack/cluster_legalizer.cpp

```diff
@@ -0,0 +1,1756 @@
+    t_mode_selection_status mode_status;
+    LegalizationCluster& cluster = legalization_clusters[cluster_id];
+    return try_intra_lb_route(cluster.router_data, log_verbosity, &mode_status);
+}
+
+ClusterLegalizer::ClusterLegalizer(const AtomNetlist& atom_netlist,
+                                   const Prepacker& t_prepacker,
+                                   const std::vector<t_logical_block_type>& logical_block_types,
+                                   std::vector<t_lb_type_rr_node>* t_lb_type_rr_graphs,
+                                   size_t t_num_models,
+                                   const std::vector<std::string>& target_external_pin_util_str,
+                                   const t_pack_high_fanout_thresholds& high_fanout_thresholds,
+                                   ClusterLegalizationStrategy t_cluster_legalization_strategy,
+                                   bool t_enable_pin_feasibility_filter,
+                                   int t_feasible_block_array_size,
+                                   int t_log_verbosity) : prepacker(t_prepacker) {
+    // Verify that the inputs are valid.
+    VTR_ASSERT_SAFE(t_lb_type_rr_graphs != nullptr);
+
+    // Resize the atom_cluster lookup to make the accesses much cheaper.
+    atom_cluster.resize(atom_netlist.blocks().size(), LegalizationClusterId::INVALID());
+    // Allocate the cluster_placement_stats
+    cluster_placement_stats = alloc_and_load_cluster_placement_stats();
+    // Pre-compute the max size of any molecule.
+    max_molecule_size = prepacker.get_max_molecule_size();
+    // Calculate the max cluster size
+    //  - Limit maximum number of elements for each cluster to MAX_SHORT
+    max_cluster_size = calc_max_cluster_size(logical_block_types);
+    VTR_ASSERT(max_cluster_size < MAX_SHORT);
+    // Get a reference to the rr graphs.
+    lb_type_rr_graphs = t_lb_type_rr_graphs;
+    // Get the number of models in the architecture.
+    num_models = t_num_models;
+    // Find all NoC router atoms.
+    std::vector<AtomBlockId> noc_atoms = find_noc_router_atoms(atom_netlist);
+    update_noc_reachability_partitions(noc_atoms,
+                                       atom_netlist,
+                                       high_fanout_thresholds,
+                                       atom_noc_grp_id);
+    // Copy the options passed by the user
+    cluster_legalization_strategy = t_cluster_legalization_strategy;
+    enable_pin_feasibility_filter = t_enable_pin_feasibility_filter;
+    feasible_block_array_size = t_feasible_block_array_size;
+    log_verbosity = t_log_verbosity;
+    // Get the target external pin utilization
+    // NOTE: This is really silly, but this can potentially fail. If it does
+    //       it is important that everything is allocated. If not, when it fails
+    //       it will call the reset method when only parts of the class are
+    //       allocated which may cause havoc...
```

Comment:
YES. This was the most painful bug to debug while working on this PR. That line of code used to be in the middle of the constructor; however I ran into the following problem:
1) The ClusterLegalizer object was created on the stack
2) The constructor was called
3) The t_ext_pin_util_targets constructor was called
4) t_ext_pin_util_targets crashed with VPR_FATAL_ERROR in the middle of it (which is expected for some testcases)
5) The ClusterLegalizer object is IMMEDIATELY destroyed (calling the destructor); since the stack is now being destroyed.
6) The destructor is called even though the object is not fully initialized causing a free of nullptr and crashing.

It was an absolute nightmare to debug and was challenging to fix in a clean way. I opted to just create the t_ext_pin_util_targets at the end of the constructor instead.

Reply:
I would mention the destructor is called if this fails (and VPR_FATAL_ERROR raises an exception) so we need to have the object in a state where the destructor will work. Hence do this last.